File: Interviewbit/SubstringConcatenation.cpp
/**
    Instead of Dynamic programming approach i sticked to hashing approach
    - Create a map for all the string in list of string array with count/frequency
    - No start iterating strings with L(length of each string) as the addition in each 
    - Maintain the count of string appread in the substring from (that start to totallength of big string) 
    - for Each iteraration see if both the hashmao match and have same frequncy of word 
    - If same add the starting index to result ArrayList
    **/

vector<int> Solution::findSubstring(string A, const vector<string> &B) {
    unordered_map<string, int> map;
    unordered_map<string, int> tempMap;
    int n = B.size();
    vector<int> result;
    if(n == 0) return result;

    int m = B[0].length();
    if(A.length() < m*n) return result;

    for(int i=0;i<n;i++)
        map[B[i]]++;

    for(int i=0;i<=A.length()-(m*n);i++){
        tempMap = map;
        bool available =false;
        for(int j=0;j<n;j++){
            int var = i+(j*m);
            string str = A.substr(var, m);
            if(tempMap.find(str) != tempMap.end()){
                if(j == n-1) available = true;
                tempMap[str]--;
                if(tempMap[str] == 0)
                    tempMap.erase(str);
            } else
                break;
        }
        
        if(available)
            result.push_back(i);
    }
    return result;
}

File: Interviewbit/SubstringConcatenation.java
/**
    Instead of Dynamic programming approach i sticked to hashing approach
    - Create a map for all the string in list of string array with count/frequency
    - No start iterating strings with L(length of each string) as the addition in each 
    - Maintain the count of string appread in the substring from (that start to totallength of big string) 
    - for Each iteraration see if both the hashmao match and have same frequncy of word 
    - If same add the starting index to result ArrayList
    **/

public class Solution {
    
    public static boolean comp(int l, String s2, Map<String, Integer> map){
        Map<String, Integer> m = new HashMap<>();
        for(int i=0;i<s2.length() -l+1;i = i+l){
            String substr = s2.substring(i, i+l);
            if(m.containsKey(substr))
                m.put(substr, m.get(substr)+1);
            else 
                m.put(substr, 1);
        }
        
        for(Map.Entry<String, Integer> x : map.entrySet())
            if(x.getValue() != m.get(x.getKey()))
                return false;
        return true;
    }
    
    public ArrayList<Integer> findSubstring(String A, final List<String> B) {
        Map<String, Integer> map1 = new HashMap<>();
        int length = B.get(0).length();
        int eachL = B.size();
        int totalLength = length*eachL;
        for(int i=0;i<B.size();i++){
            String s = B.get(i);
            if(map1.containsKey(s))
                map1.put(s, map1.get(s)+1);
            else 
                map1.put(s, 1);
        }
        ArrayList<Integer> result = new ArrayList<>();
        for(int i=0;i<A.length()-totalLength+1;i++){
            String str = A.substring(i, i+totalLength);
            if(comp(length, str, map1))
                result.add(i);
        }
        return result;
    }
}

